lib/common/string.c

```diff
@@ -352,6 +352,38 @@ h2o_iovec_t h2o_uri_escape(h2o_mem_pool_t *pool, const char *s, size_t l, const
     return encoded;
 }
 
+h2o_iovec_t h2o_uri_unescape(h2o_mem_pool_t *pool, const char *str, size_t len)
+{
+    h2o_iovec_t decoded;
+    size_t n = 0;
+    for (size_t i = 0; i < len; ++i)
+        if (str[i] == '%')
+            ++n;
+    if (n == 0)
+        return h2o_iovec_init(str, len);
+    if (2 * n >= len)
+        goto fail;
+    decoded.len = len - 2 * n;
+    decoded.base = h2o_mem_alloc_pool(pool, char, decoded.len);
+    for (size_t i = 0, j = 0; i < len; ++i, ++j) {
+        if (str[i] == '%') {
+            if (i + 2 >= len)
+                goto fail;
+            int hi = decode_hex(str[i + 1]);
+            int lo = decode_hex(str[i + 2]);
```

Comment:
Makes sense.

While I would hope that we would handle NUL characters correctly (by using vectors rather than C strings), there is probably no point in accepting NUL characters. We can start with rejecting them.

Reply:
Done in commits 8a0472b..cd2d707.